### PR TITLE
Add BFD metrics collector 

### PR DESCRIPTION
**Pull request by sabinaaledort**
The new BFD collector creates BFD metrics based on BFD peers counters.

Adding unit tests to verify BFD metrics collector functionality and extending the BFD E2E test to check BFD metrics.

### CODE REVIEW

File: e2etest/bgp_test.go [hunk 1]
```diff
@@ -534,6 +534,101 @@ var _ = ginkgo.Describe("BGP", func() {
 				return nil
 			}, 4*time.Minute, 1*time.Second).Should(BeNil())
 
+			ginkgo.By("checking metrics")
```
Comment (oribon) [inline e2etest/bgp_test.go:537]: I think the metrics tests should be a different describetable/it but under the "BFD" context.

Comment (fedepaol) [reply e2etest/bgp_test.go:626]: +1
The name must contain BFD or these tests will run in the native lanes too

File: e2etest/bgp_test.go [hunk 2]
```diff
@@ -597,6 +597,162 @@ var _ = ginkgo.Describe("BGP", func() {
...
+							err = validateCounterValue(1, "metallb_bfd_echo_packet_output", map[string]string{"peer": addr}, speakerMetrics)
+							if err != nil {
+								return err
+							}
+						}
+					}
+					return nil
+				}, 2*time.Minute, 1*time.Second).Should(BeNil())
+
+				ginkgo.By("disabling BFD in exteranl FRR containers")
```
Comment (fedepaol) [inline e2etest/bgp_test.go:689]: typo: external

File: e2etest/bgp_test.go [hunk 3]
```diff
@@ -597,6 +597,162 @@ var _ = ginkgo.Describe("BGP", func() {
...
+							return err
+						}
+
+						if bfd.EchoMode != nil && *bfd.EchoMode {
+							err = validateCounterValue(1, "metallb_bfd_echo_packet_input", map[string]string{"peer": addr}, speakerMetrics)
+							if err != nil {
+								return err
+							}
+
+							err = validateCounterValue(1, "metallb_bfd_echo_packet_output", map[string]string{"peer": addr}, speakerMetrics)
```
Comment (fedepaol) [inline e2etest/bgp_test.go:680]: Can we check also `metallb_bfd_session_down_events` is 0?

File: frr-metrics/collector/utils.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier:Apache-2.0
```
Comment (fedepaol) [inline frr-metrics/collector/utils.go:1]: Can we call this file vtysh.go? We should avoid utils and utils packages

File: e2etest/bgp_test.go (hunk 1 above)
Comment (fedepaol) [reply e2etest/bgp_test.go:626]: Ah just saw this is under BFD context.

**Comment by fedepaol** [comment]
LGTM
Let's wait CI

File: e2etest/bgp_test.go [hunk 5]
```diff
@@ -598,6 +598,172 @@ var _ = ginkgo.Describe("BGP", func() {
...
+				ginkgo.By(fmt.Sprintf("checking speaker %s", speaker.Name))
+
+				Eventually(func() error {
+					speakerMetrics, err := metrics.ForPod(controllerPod, speaker, testNameSpace)
+					if err != nil {
+						return err
+					}
+
+					for _, addr := range peerAddrs {
+						err = validateCounterValue(1, "metallb_bfd_session_down_events", map[string]string{"peer": addr}, speakerMetrics)
```
Comment (oribon) [inline e2etest/bgp_test.go:719]: Should you add a check for metallb_bfd_session_up = 0 at this point?

File: frr-metrics/collector/bfd_test.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		t.Run(test.desc, func(t *testing.T) {
+			tmpl, err := template.New(test.desc).Parse(bfdMetricsTmpl)
+			if err != nil {
+				t.Errorf("expected no error but got %s", err)
+			}
+
+			var w bytes.Buffer
+			err = tmpl.Execute(&w, map[string]interface{}{
+				"Peer":                test.peer,
+				"SessionUp":           test.sessionUp,
```
Comment (oribon) [inline frr-metrics/collector/bfd_test.go:130]: I'm not sure but I think the unit tests aren't really parsing this stat? (1 is the default)
Should we add something that covers that the collection from BFDPeers actually work?

**Comment by oribon** [comment]
LGTM

**Pull request merged**